### ISSUE TITLE
torrent_status::num_seeds is sometimes wrong

### ISSUE THREAD

**Issue by Sophist-UK** [arvidn/libtorrent#5329]
Using qBitTorrent, I find it common for connections to seeds to show < 100% progress. This is an issue of false positives reported by trackers. 

(In passing I also note that qBitTorrrent is sometimes the client, so presumably LibTorrent is also reporting a seeder with < 100% progress. This may be a question of terminology - if you have marked some files in a torrent as not to be downloaded, but you have completed the download of the rest, then you are a seeding in the sense that you are no longer downloading, but you should probably report yourself to trackers as a peer because you cannot provide the entire torrent.)

Similarly it is possible for a tracker to report a client as a Peer when it has subsequently become a seeder.

I therefore propose the following:

1. When you connect to a client in the seed list but it reports < 100% progress, you move it to the peer list and mark it as not to be tracked.

2. When you connect to a client in the peer list but it reports 100% progress, you move it to the seed list and mark it as not to be tracked.

3. When you get a tracker update, any addresses marked as not to be tracked are ignored and not moved from the lists they are in.

**Comment by arvidn** [comment]
> Using qBitTorrent, I find it common for connections to seeds to show < 100% progress. This is an issue of false positives reported by trackers.

What makes a peer a "seed" if its progress is < 100%?

Isn't a connection to a peer with progress < 100%, by definition, *not a seed*?

**Comment by arvidn** [comment]
> When you connect to a client in the peer list but it reports 100% progress, you move it to the seed list and mark it as not to be tracked.

There's only one peer list, and it include peers that have everything (i.e. seeds) and peers that don't have everything yet. What's the relevance of this "moving" you're proposing?

**Comment by Sophist-UK** [comment]
> > Using qBitTorrent, I find it common for connections to seeds to show < 100% progress. This is an issue of false positives reported by trackers.
> 
> What makes a peer a "seed" if its progress is < 100%?
> 
> Isn't a connection to a peer with progress < 100%, by definition, _not a seed_?

I haven't examined the code, but I suspect that a client defines whether it is seeding or not. For example, in qBitTorrent you can decide not to download certain files, and it will only download the chunks it needs for the remaining files. When you have downloaded all those chunks, qBitTorrent seeds the chunks it has downloaded - and I presume libTorrent reports this to trackers and possibly to other clients.

From a tracker perspective, I am unclear whether clients seeding partial downloads should report themselves as seeds or leechers. I am also unclear from a client->client perspective, whether clients seeding partial downloads should report themselves as seeds or leechers. IMO (which counts for nothing really), a client should only report to a tracker / another client that it is a seed if it is seeding a full download i.e. this should be the protocol definition of a seed - BUT of course there are many implementations of the BitTorrent protocol, and if one client changing definition to being full-downloads causes issues, it is probably difficult to change definition at this point.

So, IMO, ideally libTorrent would:

1. have a new status of "partial seeding";
2. only report to trackers / other clients that it is a seed if it is a full seed (subject to this not creating protocol issues); and
3. use the available % rather than seed status when this makes more sense (on a case by case basis)
4. display the number of full seeds to the user so that they can make better judgements about e.g. abandoning a download which has only partial seeds.

I hope this explains things a bit clearer.

**Comment by Sophist-UK** [comment]
> > When you connect to a client in the peer list but it reports 100% progress, you move it to the seed list and mark it as not to be tracked.
> 
> There's only one peer list, and it include peers that have everything (i.e. seeds) and peers that don't have everything yet. What's the relevance of this "moving" you're proposing?

I apologise for not explaining this more clearly.

libTorrent uses the information as to whether another client is a seed or a leecher to make decisions e.g. on whether to connect / disconnect.

I am suggesting that we use the % reported (for those clients we have already connected to) rather than the seed status reported either by the tracker or the client for these decisions to more accurately reflect the reality of the chunks they have available.

**Comment by arvidn** [comment]
> I am suggesting that we use the % reported (for those clients we have already connected to) rather than the seed status reported either by the tracker or the client for these decisions to more accurately reflect the reality of the chunks they have available.

The tracker does not report whether a peer is a seed or not. To the extent trackers care about this is to ensure not to hand seeds to other seeds. It sounds like your understanding of how this works is deduced from some symptoms, rather than from the protocol specification or source code.

Libtorrent's records of whether a peer is a seed or not is only affected by what the peer itself reports.

> From a tracker perspective, I am unclear whether clients seeding partial downloads should report themselves as seeds or leechers. 

This is defined in [BEP21](https://www.bittorrent.org/beps/bep_0021.html).


>     have a new status of "partial seeding";

It already does. It's called "upload_only".

>     only report to trackers / other clients that it is a seed if it is a full seed (subject to this not creating protocol issues); and

There's no way to report oneself as a seed to anyone. You report your progress. Being a seed is just a name for the case where you have 100% of the file(s). However, you can report yourself as upload-only (see BEP21).

>     use the available % rather than seed status when this makes more sense (on a case by case basis)

Those are the same things.

>     display the number of full seeds to the user so that they can make better judgements about e.g. abandoning a download which has only partial seeds.

I think this is where the meat of your proposal is. Are you looking at tracker scrape information and are disappointed when you can't connect to the same number of seeds?

There are lots of reasons why you may be connected to a different number of seeds than the tracker reports in its scrape information.

libtorrent also reports the number of seeds in its peer-list. This is slightly more reliable because it represents peers we've actually been connected to and they've said they have everything. There's no guarantee that they are still around though, or that they accept incoming connections.

So, is this thread just a matter of reporting numbers to users?

**Comment by Sophist-UK** [comment]
You are correct - I didn't have time to read and understand the BitTorrent protocol spec or to read the code to see what LibTorrent does. I am reporting the symptoms as seen in qBitTorrent on the assumption that qBitTorrent is displaying what is passed to it by libTorrent, and this could of course be qBitTorrent mis-classifying what it is told by libTorrent.

I sometimes get qBitTorrent showing torrents with multiple seeds and no leechers yet all the seeds only have a few % of the torrent and when connected they are shown as connected as a seed in the 1-line torrent summary, but when you look at the list of connections for the torrent you can see that they don't have 100% of the torrent.

**Comment by arvidn** [comment]
> I sometimes get qBitTorrent showing torrents with multiple seeds and no leechers yet all the seeds only have a few % of the torrent 

I don't understand what this means. When you say "all seeds only have a few % of the torrent". What makes them seeds? I would think they would *not* be seeds, by definition. Does qBT mark individual peers as seeds or not in its peer list?

**Comment by arvidn** [comment]
oh, you're talking about a counter of number of "connected seeds". Is that right?

**Comment by Sophist-UK** [comment]
No - I have now checked the qBitTorrent code and it appears that when it checks `isSeed` it doesn't check `isUploadOnly`.

I will submit a PR to qBitTorrent. Thanks for your patience and help.

**Comment by Sophist-UK** [comment]
Or maybe I didn't understand BEP21.

Are partial seeds shown by libTorrent as `seed` **and** `upload_only` or just `upload_only`?

**Comment by arvidn** [comment]
both. the `upload_only` and `seed` flags are set. http://libtorrent.org/reference-Core.html#flags

**Comment by Sophist-UK** [comment]
Great - my PR should be correct in that case.

**Comment by Sophist-UK** [comment]
https://github.com/qbittorrent/qBittorrent/pull/14508

**Comment by Sophist-UK** [comment]
I am confused now - as the comments you made on the PR seem to contradict what you stated above.

**Comment by arvidn** [comment]
oh, I see. You asked which flags are set for "partial seeds" (i.e. incomplete upload-only peers). then `upload_only` is set, and not `seed`.

The documentation is [here](http://libtorrent.org/reference-Core.html#seed) and [here](http://libtorrent.org/reference-Core.html#upload_only).

**Comment by arvidn** [comment]
let's avoid the term "partial seed" :)

**Comment by Sophist-UK** [comment]
Ok - so, now I understand the protocol and how libTorrent handles things, I need to look elsewhere for an explanation of what is happening and elsewhere for a fix.

As I understand it libTorrent gets seed statistics in two ways:

a. From tracker scraping - but it only gets a summary number and although the tracker gives the address and port of all the peers, it doesn't say whether it thinks they are seeders or leechers.

b. From the individual peers when you connect to them and they provide their progress. I assume that progress=1 / progress_ppm=1,000,000 are what defines an individual peer as being a seed.

But we also have a bitmap of pieces that defines whether the peer has all the pieces, which is another way of determining whether the peer is a seed or not.

So one potential explanation (the only explanation?) is that some peers are reporting progress = 1 (which makes them a peer), but when you actually try to download they don't have all the pieces. I am unclear whether these peers are reporting a full bitmap and then pieces just fail when you try to download them, or that bitmap is correct for the pieces they actually have but the progress is incorrectly calculated.

I guess I need to look at the qBitTorrent code again and see how the connection % is calculated - if it is from the bitmap, then we have an explanation as to the mismatch - but if it is using progress then I am flummoxed.

**Comment by arvidn** [comment]
> But we also have a bitmap of pieces that defines whether the peer has all the pieces, which is another way of determining whether the peer is a seed or not.

We do have that bitfield, but we don't *also* have the bitfield, that *is* how peers report their progress.

> So one potential explanation (the only explanation?) is that some peers are reporting progress = 1 (which makes them a peer)

You being connected to them and speaking the bittorrent protocol makes them a peer, regardless of their download progress.

> , but when you actually try to download they don't have all the pieces. I am unclear whether these peers are reporting a full bitmap and then pieces just fail when you try to download them, or that bitmap is correct for the pieces they actually have but the progress is incorrectly calculated.

What is the symptom you experience that this would be an explanation to?
The bittorrent protocol only has one way for peers to communicate their progress, by saying which pieces they have. i.e. the bitfield. This bitfield is updated over time though, by `HAVE` messages.

> I guess I need to look at the qBitTorrent code again and see how the connection % is calculated - if it is from the bitmap, then we have an explanation as to the mismatch - but if it is using progress then I am flummoxed.

Those fields represent the same thing, one is just a summary of the other.

I still don't understand what the problem is. You seem to assume that peers are considered seeds somewhere, even though they report < 100% completion. Where do you get that impression from?

Is the issue that something is telling you that you are connected to x number of seeds, and in the peer list you don't see any seeds?

**Comment by Sophist-UK** [comment]
I start to download a torrent. It has (say) 4 seeds and no peers.

But when the seeds are connected, qBitTorrent shows each of them has (say) 99.2% piece availability. And you only get 99.2% downloaded yourself before the download stalls. And one or more files are incomplete. So in reality, there are no genuine seeds, even though several were shown as existing in qBitTorrent. I am assuming that qBT gets its information about seeds from libTorrent, though as previously discussed it could be that libToprrent has it right and QBT is misrepresenting what libTorrent has   told it.

Clearly QBT has associated the specific connections as seeds because the connections are shown in the seeds column. So it doesn't seem to be something to do with tracker summary statistics about seeds.

So something appears to be defining these particular IP addresses as seeds even though they aren't. We seem to have eliminated trackers (that don't provide information about  specific peers) and the peer itself (because it only shares the bitfield as the means of showing whether it is a seed or not and that is the same as progress).

I am now wondering what other parts of the extended bittorrent protocol could cause this:

* PEX - could it be that the seed/upload_only bit is being treated as seed?
* Fast Extensions - HAVE_ALL bit could be set when client doesn't have all the pieces - I guess this should be used for scheduling etc. but perhaps not to identify peer as a seed

Obviously I have only limited knowledge of bittorrent protocols and libTorrent, but this is a real issue and it must have a cause somewhere.

**Comment by arvidn** [comment]
> I start to download a torrent. It has (say) 4 seeds and no peers.

How do you know this in advance? Also, I think it's risky to use the term "peer" to mean a bittorrent client that does not have all the files. A peer is a well established term in peer-to-peer systems, and seeds are also peers. You mean it shows no incomplete peers.

> But when the seeds are connected

How do you know the peers you connected to are seeds?

> qBitTorrent shows each of them has (say) 99.2% piece availability.

Because it sounds like they aren't.

> So in reality, there are no genuine seeds, even though several were shown as existing in qBitTorrent

Meaning, the tracker scrape information said there were seeds, right?

> I am assuming that qBT gets its information about seeds from libTorrent, though as previously discussed it could be that libToprrent has it right and QBT is misrepresenting what libTorrent has told it.

But libtorrent gets scrape information from the tracker, so (probably more likely) the tracker is wrong. I would expect scrape information to possibly be outdated. Alternatively, all the seeds are behind NAT and cannot be connected. Perhaps the incomplete peers you connected to came from the DHT, which could explain the tracker not knowing about them.

> Clearly QBT has associated the specific connections as seeds because the connections are shown in the seeds column. So it doesn't seem to be something to do with tracker summary statistics about seeds.

I'm unfamiliar with this. Are peer connections listed in different columns based on whether they are seeds or not?
Being shown in the "seeds" column, is that based on the `seed` flag being set on the `peer_info` object?

The logic determining whether the `seed` flag is set or not can be found [here](https://github.com/arvidn/libtorrent/blob/RC_2_0/src/peer_connection.cpp#L6766).

> So something appears to be defining these particular IP addresses as seeds even though they aren't

It would be helpful to understand what criteria is used to determine whether a peer connection should be considered a seed or not then, in the qBT UI.

**Comment by arvidn** [comment]
here's a (low resolution) screenshot of the qBT peer list. I only see a single column of peers. What property are you looking at to consider the peers "seeds" when they report < 100% completion?

https://a.fsdn.com/con/app/proj/qbittorrent/screenshots/242193.jpg

**Comment by Seeker2** [comment]
I've got a seeding torrent that regularly sees peers-reported-as-seeds by qBitTorrent:
![qBitTorrent peers-as-seeds](https://user-images.githubusercontent.com/4667526/110381910-c3aa6e00-801f-11eb-8028-25583dbf4dea.png)

There is clearly upload activity to them, and the last one on the list disconnected and dropped the number of seeds down to 2 shortly after the picture was taken (may already be DCed, due to the speed cool-down not dropping instantly to 0)
...so it may not even be the 90+% complete ones that are mis-identified this way.

**Comment by arvidn** [comment]
in that screenshot I see 5 peers, none of which is a seed. In what way are they reported as seeds?

**Comment by Seeker2** [comment]
The highlighted torrent in the top part of the screenshot is reporting Seeds 3 (14) and Peers 2 (9).

**Comment by Sophist-UK** [comment]
So there are 5 connections shown in the torrent summary (3 seeds, 2 leechers) but when you look at the completeness you can see that all 5 are incomplete and uploaders. How can a "seed" be both incomplete AND an uploader?

**Comment by Sophist-UK** [comment]
This is a similar example to what I experienced **except** that I was a downloader not a seeder, and I was seeing seed connections which were incomplete.

**Comment by Sophist-UK** [comment]
> here's a (low resolution) screenshot of the qBT peer list. I only see a single column of peers. What property are you looking at to consider the peers "seeds" when they report < 100% completion?
> 
> https://a.fsdn.com/con/app/proj/qbittorrent/screenshots/242193.jpg

@arvidn - You cannot use a normal torrent with genuine seeds as an example. We all know that the majority of torrents are shown correctly. I am reporting a special case - which affect a minority of torrents but nevertheless a multitude - where something is causing incomplete peers to appear as seeds - and they may either still be attempting to download but are appearing to me as seeds, or they may be incomplete upload_only peers appearing to me as seeds.

I have no idea why - I just know that this happens. And because it has seeds, it causes me to believe that the torrent can be downloaded in full when in fact there are NO complete peers at all, and so no genuine seeds, and it will never download completely.

**Comment by Sophist-UK** [comment]
And as ever, now that I want to find an example in order to take a screenshot, I cannot now do so. {sigh}

**Comment by arvidn** [comment]
> So there are 5 connections shown in the torrent summary (3 seeds, 2 leechers)

Ok, it's the numbers that are the problem. The ones before the parenthesis. What numbers are those? where do they come from?

> but when you look at the completeness you can see that all 5 are incomplete and uploaders. How can a "seed" be both incomplete AND an uploader?

I don't think any of the peers you see in the peer list are seeds.

**Comment by arvidn** [comment]
> You cannot use a normal torrent with genuine seeds as an example.

I was trying to understand what you meant by the "seed" column. But I get it now, you weren't talking about peer connections, you were talking about the counters in the torrent list.

**Comment by arvidn** [comment]
> The highlighted torrent in the top part of the screenshot is reporting Seeds 3 (14) and Peers 2 (9).

I take it "3" is supposed to mean there are 3 connected seeds, and "2" means there are 2 connected incomplete peers (I think it's weird to only call incomplete peers "peers").

So, where do those numbers come from? It seems they are wrong and the question is, is that because of a bug or is it because they just don't mean what people think they mean.

What about the numbers in parenthesis, what are they?

**Comment by Seeker2** [comment]
> 
> > So there are 5 connections shown in the torrent summary (3 seeds, 2 leechers)
> 
> Ok, it's the numbers that are the problem. The ones before the parenthesis. What numbers are those? where do they come from?

The number before the parenthesis is currently connected peers (or seeds, as the case may be).
The number in parenthesis is reported numbers from trackers, PEX, DHT, LSP/LPD, and added manually (assuming these are used).
I had DHT and LSP/LPD disabled and wasn't adding any peers/seeds manually.
The torrent has a handful of working trackers in 2 tracker tiers, which it only uses 1 tracker per tier...and I check maybe once-a-week that the trackers are active.
PEX plus qBT not being firewalled keeps the torrent swarm "knit together".

**Comment by arvidn** [comment]
So, these field appears to be wrong, yes? https://libtorrent.org/reference-Torrent_Status.html#num_peers https://libtorrent.org/reference-Torrent_Status.html#num_seeds
Assuming those are the fields used to populate those columns

**Comment by Seeker2** [comment]
> So, where do those numbers come from? It seems they are wrong and the question is, is that because of a bug or is it because they just don't mean what people think they mean.

As for "is this a bug"?

Yes, it is... uTorrent on the same torrent never has this problem, even when I run it on a separate computer on a different internet connection. It often connects to a few extra peers.

**Comment by arvidn** [comment]
> As for "is this a bug"?
> 
> Yes, it is... uTorrent on the same torrent never has this problem, even when I run it on a separate computer on a different internet connection. It often connects to a few extra peers.

Let me rephrase: "is this a bug in libtorrent?"

**Comment by Seeker2** [comment]
From everything I've seen, yes the problem is likely in libtorrent.

**Comment by thalieht** [comment]
>What numbers are those? where do they come from?

Outside the parentheses is `num_seeds`
Inside is 
```
(num_complete > 0) ? num_complete : list_seeds;
```

**Comment by arvidn** [comment]
looking into this, it seems there's some logic marking upload-only peers as seeds in the peer-list. Presumably this was to leverage the existing logic to avoid connecting to seeds when one self is seed or upload-only. However, the logic looks suspicious as there doesn't seem to be a way to get *unmarked* as a seed in upload-only mode is disabled. It also affects the accounting, specifically the `num_seeds` counter.

It seems reasonable to treat upload-only as a more ephemeral state, and not record it in the peer-list, but just in the peer-connection.

**Comment by Seeker2** [comment]
In my screenshot, I was uploading to all the peers.
I doubt they were upload-only peers.

**Comment by arvidn** [comment]
of course it's possible what I've found isn't the cause of this issue. However, a peer that was upload-only at some point, and then changed to be downloading, currently it wouldn't be unmarked as a seed. According to [this](https://github.com/arvidn/libtorrent/blob/RC_1_2/src/peer_connection.cpp#L6630) logic.

that `||` is most likely supposed to be a `&&`.

**Comment by Seeker2** [comment]
Someone may have used file-level priority to cause weirdness, like "don't download" then changed them later.

**Comment by Sophist-UK** [comment]
It seems to me that the following transitions are possible:

a. Downloader to Seeder - normal when all pieces are finally downloaded
b. Seeder to downloader - user deletes a file and so it needs to be downloaded again - probably needs a Pause / Check / Resume, but if you are not connected to the peer at the time this might all happen in the gap.
c. Downloader to upload_only - normal when you have decided not to download one or more files.
d. upload_only to downloader - normal when you decide to download a file you previously decided not to.
e. upload_only to seeder - unsure how this could happen - but I suppose it might occur if you were previously a seeder, then marked a file as not to download and then change your mind (but since you have all the pieces anyway, I would imagine that you should have remained a seeder).
f. seeder to upload_only - see b. and e.

**Comment by arvidn** [comment]
I think this patch may address it: https://github.com/arvidn/libtorrent/pull/6054

**Comment by Sophist-UK** [comment]
Screenshort from qbT 4.3.5 containing libtorrent-rasterbar 1.2.13+git22af326f6c:
![image](https://user-images.githubusercontent.com/3001893/121344404-a2f1fd00-c91b-11eb-9c82-21d6ee0dfec7.png)
As you can see, I have several torrents with persistent connections to seeds with no pieces I am interested in. IMO staying connected to these peers adds zero value, and libTorrent should disconnect from them and connect again after a defined period. This period might be measured in minutes - possibly even an increasing interval starting at say 1 minute and increasing to say 10-minutes if previous connections still do not have pieces of interest.

**Comment by Seeker2** [comment]
They at least appear to be peers. If they have the same percentage complete as you, staying connected means your peer will likely get early access to any pieces they download.
(I'm guessing an occasional slow seed is connecting on these torrents?)

**Comment by Sophist-UK** [comment]
Hmmm - I think I may have added this to the wrong issue.

**Comment by Sophist-UK** [comment]
Nudge to remove staleness.
